Dotty internals docs update, dotty knowledge collection system integration (#6786)
lampepfl/dotty: docs/docs/contributing/debugging.md

- [Show for human readable output](#show-for-human-readable-output)
- [How to disable color](#how-to-disable-color)
- [Reporting as a non-intrusive println](#reporting-as-a-non-intrusive-println)
- [Printing out trees after phases](#printing-out-trees-after-phases)
- [Configuring the printer output](#configuring-the-printer-output)
- [Figuring out an object creation site](#figuring-out-an-object-creation-site)
  * [Via ID](#via-id)
  * [Via tracer](#via-tracer)
- [Built-in Logging Architecture](#built-in-logging-architecture)
  * [Printers](#printers)
  * [Tracing](#tracing)
  * [Reporter](#reporter)

<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>

## Setting up the playground
Consider the `../issues/Playground.scala` (relative to the Dotty directory) file is:

```scala
object Playground {
  def main(args: Array[String]) = {
    println("Hello World")
  }
}
```

Then, you can debug Dotty by compiling this file via `dotc ../issues/Playground.scala` (from the SBT console) and collecting various debug output in process. This section documents techniques you can use to collect the debug info.

> [This](https://github.com/lampepfl/dotty/blob/10526a7d0aa8910729b6036ee51942e05b71abf6/compiler/src/dotty/tools/dotc/typer/Typer.scala#L2231) is the entry point to the Typer. The job of the Typer is to take an untyped tree, compute its type and turn it into a typed tree by attaching the type information to that tree (in an immutable way of course). We will use this entry point to practice debugging techniques. E.g.:

Reviewer: > (in an immutable way of course)

Technically, we do use mutation to set types on trees: https://github.com/lampepfl/dotty/blob/dc0f865988b2abe84345bf63213087c8af5809f2/compiler/src/dotty/tools/dotc/ast/Trees.scala#L73-L79, so I would remove this comment.

val YtestPickler      = BooleanSetting("-Ytest-pickler"      , "self-test for pickling functionality; should be used with -Ystop-after:pickler")
```

They are defined in [ScalaSettings.scala](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/config/ScalaSettings.scala). E.g. `YprintPos` is defined as:

```scala
val YprintPos: Setting[Boolean] = BooleanSetting("-Yprint-pos", "show tree positions.")
```

And is to be used as:

```scala
dotc -Yprint-pos  ../issues/Playground.scala
```

If used, all the trees output with `show` or via `-Xprint:frontend` will also have positions attached to them, e.g.:

```scala
package <empty>@<Playground.scala:1> {
  module object Playground {
    def main(
      args:
        Array@<Playground.scala:2>[String@<Playground.scala:2>]@<
          Playground.scala:2
        >
      @<Playground.scala:2>
    ) =
      {
        println@<Playground.scala:3>("Hello World"@<Playground.scala:3>)@<
          Playground.scala:3
        >
      }@<Playground.scala:2>
    @<Playground.scala:2>
  }@<Playground.scala:1>
}@<Playground.scala:1>
<empty>@<Playground.scala:1>
```

## Figuring out an object creation site
### Via ID

> Every [Positioned](https://github.com/lampepfl/dotty/blob/10526a7d0aa8910729b6036ee51942e05b71abf6/compiler/src/dotty/tools/dotc/ast/Positioned.scala) (a trait of a `Tree`) object has `uniqueId` field. It is an integer that is unique for that tree and doesn't change from compile run to compile run. You can output these IDs from the printers (= from `show` and `-Xprint`) framework via `-Yshow-tree-ids` flag, e.g.:

Reviewer: ```suggestion
Every [Positioned](https://github.com/lampepfl/dotty/blob/10526a7d0aa8910729b6036ee51942e05b71abf6/compiler/src/dotty/tools/dotc/ast/Positioned.scala) (a parent class of `Tree`) object has a `uniqueId` field. It is an integer that is unique for that tree and doesn't change from compile run to compile run. You can output these IDs from any printer (such as the ones used by `.show` and `-Xprint`) via `-Yshow-tree-ids` flag, e.g.:
```

Author: Maybe "a parent trait" would be better?

Reviewer: It's a class, not a trait.

Author: Oh, indeed :)

```
dotc -Xprint:frontend -Yshow-tree-ids  ../issues/Playground.scala
```

Gives:

```scala
package <empty>#1047 {
  final lazy module val Playground: Playground$#1049 =
    new Playground$#1049#1050#1051()#1052
  #1053
  final module class Playground$() extends Object#1090#1091#1092()#1093, _root_#
    1061
  .scala#1062.Serializable#1063 { this: Playground#1054.type#1055 =>
    def main(args: Array#1028[String#1033]#1034#1038): Unit#1039 =
      {
        println#1094("Hello World"#1041)#1095
      }#1096
    #1097
  }#1099
}#1100
```

You can then use these IDs to locate the creation site of a given tree using that ID via `-Ydebug-tree-with-id`, e.g.:

```
dotc -Ydebug-tree-with-id 1049 ../issues/Playground.scala
```

> When requested what you want to do:

Reviewer: ```suggestion
When the tree with the correspond id is allocated, the following prompt will appear:
```

```
Debug tree (id=1049) creation
Ident(Playground$)

a)bort, s)tack, r)esume
```

> Input `s` for stack trace. You will get:

Reviewer: ```suggestion
If you input `s`, you will get a stack trace like this:
```

  at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:102)
  at dotty.tools.dotc.Run.compileUnits(Run.scala:185)
  at dotty.tools.dotc.Run.compileSources(Run.scala:120)
  at dotty.tools.dotc.Run.compile(Run.scala:104)
  at dotty.tools.dotc.Driver.doCompile(Driver.scala:34)
  at dotty.tools.dotc.Driver.process(Driver.scala:172)
  at dotty.tools.dotc.Driver.process(Driver.scala:141)
  at dotty.tools.dotc.Driver.process(Driver.scala:153)
  at dotty.tools.dotc.Driver.main(Driver.scala:180)
  at dotty.tools.dotc.Main.main(Main.scala)
```

So that tree was created at:

```
  at dotty.tools.dotc.ast.desugar$.moduleDef(Desugar.scala:804)
```

Since all the stack frames above it are technical frames executing the tree creation command, and the frame in question is the location where the intent of the tree creation was expressed.

### Via tracer
Some objects may not be `Positioned` and hence their creation site is not debuggable via the technique in the section above. Say you target a tree at `Typer`'s `typed` method as follows:

```scala
if (tree.show == """println("Hello World")""") {
  val creationSite = "<creation site stack here>"
  println(creationSite)
}
```

> In other words, you have a reference to the object and want to know were it was created. To do so, go to the class definition of that object. In our case, `tree` is a [`Tree`](https://github.com/lampepfl/dotty/blob/10526a7d0aa8910729b6036ee51942e05b71abf6/compiler/src/dotty/tools/dotc/ast/Trees.scala#L52). Now, create a new `val` member of that type:

Reviewer: It's a bit weird to use this technique on a tree, when trees already have `-Ydebug-tree-with-id`, maybe use a different type for the example ?

Author: True. I've logged it at lampepfl/dotty-knowledge#6, will address with other doc issues later.

```scala
val tracer = Thread.currentThread.getStackTrace.mkString("\n")
```

Then, from the `typed` method (or wherever you want to access the trace):

```scala
if (tree.show == """println("Hello World")""") {
  val creationSite = tree.tracer
  println(creationSite)
}
```

## Built-in Logging Architecture

> Dotty has a lot of debug calls scattered throughout the code, vast majority of them disabled. At least three (possibly intertwined) architectures for logging are used for that:

Reviewer: ```suggestion
Dotty has a lot of debug calls scattered throughout the code, most of which are disabled by default. At least three (possibly intertwined) architectures for logging are used for that:
```
